ui/src/app/machines/views/MachineDetails/MachineTests/MachineTests.tsx

```diff
@@ -4,13 +4,61 @@ import { Spinner } from "@canonical/react-components";
 import { useDispatch, useSelector } from "react-redux";
 import { useParams } from "react-router";
 
+import MachineTestsTable from "./MachineTestsTable";
+
+import { HardwareType, ResultType } from "app/base/enum";
 import { useWindowTitle } from "app/base/hooks";
 import type { RouteParams } from "app/base/types";
 import machineSelectors from "app/store/machine/selectors";
 import { actions as nodeResultActions } from "app/store/noderesult";
 import nodeResultSelectors from "app/store/noderesult/selectors";
+import type { NodeResult, NodeResults } from "app/store/noderesult/types";
 import type { RootState } from "app/store/root/types";
 
+/**
+ * Group items by key
+ * @param results a node results list
+ * @param key
+ */
+const groupByKey = (items, key: string) =>
+  items.reduce((obj, item) => {
+    obj[item[key]] = obj[item[key]] || [];
+    obj[item[key]].push(item);
+    return obj;
+  }, Object.create(null));
+
+const getTestingResults = (nodeResults: NodeResults) =>
```

Comment:
Considering moving these into selectors, but for the purposes of an initial branch, hopefully this is okay.